mobile_packet_verifier/src/accumulate.rs

```diff
@@ -60,3 +77,55 @@ pub async fn accumulate_sessions(
 
     Ok(())
 }
+
+async fn verify_report(
+    gateway_client: &GatewayClient,
+    auth_client: &AuthorizationClient,
+    report: &DataTransferSessionReq,
+) -> DataTransferIngestReportStatus {
+    if !verify_gateway(gateway_client, &report.data_transfer_usage.pub_key).await {
+        return DataTransferIngestReportStatus::InvalidGatewayKey;
+    };
+    if !verify_known_routing_key(auth_client, &report.pub_key).await {
+        return DataTransferIngestReportStatus::InvalidRoutingKey;
+    };
+    DataTransferIngestReportStatus::Valid
+}
+
+async fn verify_gateway(gateway_client: &GatewayClient, public_key: &PublicKeyBinary) -> bool {
+    match gateway_client.resolve_gateway_info(public_key).await {
+        Ok(res) => res.is_some(),
+        Err(_err) => false,
+    }
+}
+
+async fn verify_known_routing_key(
+    auth_client: &AuthorizationClient,
+    public_key: &PublicKeyBinary,
+) -> bool {
+    match auth_client
+        .verify_authorized_key(public_key, NetworkKeyRole::MobileRouter)
+        .await
+    {
+        Ok(res) => res,
+        Err(_err) => false,
+    }
+}
+
+async fn write_invalid_report(
+    invalid_data_session_report_sink: &FileSinkClient,
+    reason: DataTransferIngestReportStatus,
+    report: DataTransferSessionIngestReport,
+) -> Result<(), file_store::Error> {
+    let proto: InvalidDataTransferIngestReportV1 = InvalidDataTransferIngestReport {
```

Comment:
this seems like an odd way to construct this struct; why not just initialize it directly?

Reply:
the long running convention ( as i understand it ) has been to only construct the internal representations of the structs and then convert to the proto version.